File: tests/FXRateDashboard.Tests/RateQueryMapperTests.cs
using FXRateDashboard.Models;
using FXRateDashboard.Services;

namespace FXRateDashboard.Tests;

public sealed class RateQueryMapperTests
{
    [Theory]
    [InlineData(TimeRangePreset.Day, "minute", -1)]
    [InlineData(TimeRangePreset.Week, "hour", -7)]
    [InlineData(TimeRangePreset.Month, "day", -30)]
    [InlineData(TimeRangePreset.Year, "day", -365)]
    public void MapHistoryQuery_ReturnsExpectedWindow(TimeRangePreset range, string expectedGroup, int dayDelta)
    {
        var mapper = new RateQueryMapper();
        var anchor = new DateTimeOffset(2026, 4, 10, 8, 30, 0, TimeSpan.Zero);

        var query = mapper.MapHistoryQuery(range, anchor);

        Assert.Equal(expectedGroup, query.Group);
        Assert.Equal(anchor, query.ToUtc);
        Assert.Equal(anchor.AddDays(dayDelta), query.FromUtc);
    }
}

File: macos/FXRateDashboard/Tests/FXRateDashboardKitTests/RateQueryMapperTests.swift
import Foundation
import XCTest
@testable import FXRateDashboardKit

final class RateQueryMapperTests: XCTestCase {
    func testMapHistoryQueryReturnsExpectedWindows() {
        let cases: [(TimeRangePreset, String, TimeInterval)] = [
            (.day, "minute", -86_400),
            (.week, "hour", -604_800),
            (.month, "day", -2_592_000),
            (.year, "day", -31_536_000)
        ]

        let mapper = RateQueryMapper()
        let anchor = Date(timeIntervalSince1970: 1_744_360_200)

        for (range, group, delta) in cases {
            let query = mapper.mapHistoryQuery(range, nowUTC: anchor)
            XCTAssertEqual(query.group, group)
            XCTAssertEqual(query.toUTC, anchor)
            XCTAssertEqual(query.fromUTC, anchor.addingTimeInterval(delta))
        }
    }
}
